katsdpimageutils/primary_beam_correction.py

```diff
@@ -6,269 +6,305 @@
-    ----------
-    path : str
-        FITS file
-
-    Returns
-    -------
-    output_file : astropy.io.fits.hdu.image.PrimaryHDU
-        First element of the HDU list
-    """
-    fl = fits.open(path)
-    images = fl[0]
-    return images
-
-
-def get_position(path):
-    """Determine the sky coordinate of the pointing centre.
-
-    This implementation assumes that the pointing centre is the
-    same as the crval.
-
-    Parameters
-    ----------
-    path : str
-        FITS file
-
-    Returns
-    -------
-    position : astropy.coordinations.SkyCoord
-        Sky coordinate of the phase centre
-    image_wcs : astropy.wcs.wcs.WCS
-        WCS keywords in the primary HDU
-    """
-    # Parse the WCS keywords in the primary HDU
-    image_wcs = wcs.WCS(path)
-    # Get pointing centre of the observation
-    phase_centre_ra = image_wcs.celestial.wcs.crval[0]
-    phase_centre_dec = image_wcs.celestial.wcs.crval[1]
-    # Convert to astropy.coordinates.SkyCoord object
-    phase_center = coordinates.SkyCoord(phase_centre_ra, phase_centre_dec, unit=(u.deg, u.deg))
-    return phase_center, image_wcs
 
+_cosine_taper = katbeam.jimbeam._cosine_taper
 
-def radial_offset(phase_center, image_wcs):
-    """Compute radial offset of pixels from the phase centre in radians.
 
-    Parameters
-    ----------
-    phase_center : astropy.coordinations.SkyCoord
-        phase center position
-    image_wcs : astropy.wcs.wcs.WCS
-        WCS keywords in the primary HDU
-    """
-    # Get pixel coordinates
-    pixcrd = np.indices((image_wcs.array_shape[2], image_wcs.array_shape[3]))
-    row = np.ravel(pixcrd[0])
-    col = np.ravel(pixcrd[1])
-    # Convert pixel coordinates to world coordinates
-    p2w = image_wcs.pixel_to_world(col, row, 0, 0)[0]
-    # Compute a separation vector between phase centre and source positions in radians.
-    separation_rad = p2w.separation(phase_center).rad
-    return separation_rad
-
-
-def central_freq(path):
-    """Determine central frequency of each frequency plane.
+def _circular_pattern(x, y, fwhm_x, fwhm_y):
+    """Make the beam circular.
 
     Parameters
     ----------
-    path : str
-        FITS file
-
-    Returns
-    -------
-    output : numpy array
-        An array of central frequencie in MHz of each frequency plane.
+    x, y : arrays of float of the same shape
+       Coordinates where beam is sampled, in degrees
     """
-    images = read_fits(path)
-    c_freq_plane = []
-    # NSPEC -  number of frequency planes.
-    for i in range(1, images.header['NSPEC']+1):
-        # FREQ00X is the central frequency for each plane X.
-        c_freq_plane.append(images.header['FREQ{0:04}'.format(i)])
-    return np.array(c_freq_plane)
+    theta_b = np.sqrt(fwhm_x * fwhm_y)
+    r = np.sqrt(x**2 + y**2)
+    rr = r/theta_b
+    return _cosine_taper(rr)
 
 
-def cosine_power_pattern(separation_rad, c_freq):
-    """Compute Power patterns for a given frequency.
+class Circular_Beam(JimBeam):
```

Comment:
I'm not sure this is the best way to do this.

A less drastic change might be to just override the `_pattern` function. Something like:

```
import katbeam

def my_pattern(x, y, squint_x, squint_y, fwhm_x, fwhm_y):
    return _circular_pattern(x, y, fwhm_x, fwhm_y)

katbeam._jimbeam._pattern = my_pattern

my_beam = katbeam.JimBeam(...)
thebeam = my_beam(x,y,freq)
```

This works for me in an ipython session.

Reply:
Actually after discussion its probably better to have your `CircleBeam` inherit `JimBeam` and override the HH and VV methods than what I've suggested here.

The issue is that you shouldn't make all the module functions into class methods, `CircleBeam` should just have HH and VV methods (as well as those inherited from `JimBeam`